---
title: "Scrublet"
author: "Wenjun Zhu"
date: "10/24/2022"
output: pdf_document
---

```{r setup, include=FALSE}
library(knitr)
library(reticulate)
knitr::knit_engines$set(python = reticulate::eng_python)
```


### Setup Python in R

To use Python in R, we need to load the reticulat R package.

```{r}
library(reticulate)
```


### Load libraries

```{python libraries}
import scrublet as scr
import scipy.io
import matplotlib.pyplot as plt
import numpy as np
import os
```


### Customize plot elements

```{python customize plt}
plt.rcParams['font.family'] = 'sans-serif'
plt.rcParams['font.sans-serif'] = 'Arial'
plt.rc('font', size=14)
plt.rcParams['pdf.fonttype'] = 42
```


### Load counts matrix and gene list

Load the raw counts matrix as a scipy sparse matrix with cells as rows and genes as columns.

```{python data}
input_dir = '/Users/jasmine/Documents/GNS/5cdc540d328cee7a2efc234a/'
counts_matrix = scipy.io.mmread(input_dir + 'matrix.mtx').T.tocsc()
genes = np.array(scr.load_genes(input_dir + 'genes.tsv'))
cells = np.array(scr.load_genes(input_dir + 'barcodes2.tsv'))

print('Counts matrix shape: {} rows, {} columns'.format(counts_matrix.shape[0], counts_matrix.shape[1]))
print('Number of genes in gene list: {}'.format(len(genes)))
```


### Initialize Scrublet object

The relevant parameters are:
1. expected_doublet_rate: the expected fraction of transcriptomes that are doublets, typically 0.05-0.1.
2. sim_doublet_ratio: the number of doublets to simulate, relative to the number of observed transcriptomes. The default value is 2.
3. n_neighbors: Number of neighbors used to construct the KNN classifier of observed transcriptomes and simulated doublets. The default value is round(0.5*sqrt(n_cells)).

```{python Scrublet object}
scrub = scr.Scrublet(counts_matrix, expected_doublet_rate=0.01)
```


### Run the default pipeline

1. Doublet simulation
2. Normalization, gene filtering, rescaling, PCA
3. Doublet score calculation
4. Doublet score threshold detection and doublet calling

```{python Scrublet pipeline}
doublet_scores, predicted_doublets = scrub.scrub_doublets(min_counts=50, # gene counts
                                                          min_cells=50,  # cell number
                                                          min_gene_variability_pctl=50, 
                                                          log_transform = True,
                                                          n_prin_comps=15)  # of principle components 
```


### Plot doublet score histograms for observed transcriptomes and simulated doublets

The simulated doublet histogram is typically bimodal. The left mode corresponds to "embedded" doublets generated by two cells with similar gene expression. The right mode corresponds to "neotypic" doublets, which are generated by cells with distinct gene expression (e.g., different cell types) and are expected to introduce more artifacts in downstream analyses. Scrublet can only detect neotypic doublets.

To call doublets vs. singlets, we must set a threshold doublet score, ideally at the minimum between the two modes of the simulated doublet histogram. Or we can adjust the threshold with the call_doublets() function. For example:

scrub.call_doublets(threshold=0.75)

```{python Scrublet plot doublet score histogram}
scrub.call_doublets(threshold=0.25)
scrub.plot_histogram()
plt.show()
```


### Get 2-D embedding to visualize the results

```{python Scrublet 2-D embedding}
print('Running UMAP...')
scrub.set_embedding('UMAP', scr.get_umap(scrub.manifold_obs_, 10, min_dist=0.3))
print('Done.')
```


### Plot doublet predictions on 2-D embedding

Predicted doublets should co-localize in distinct states.

```{python Scrublet plot doublet}
scrub.plot_embedding('UMAP', order_points=True)
plt.show()
```


### Type in 'exit' or 'quit' to exit the REPL and return to R.
